### ISSUE TITLE
Unnecessary use of a comprehension: Wrong fix

### ISSUE THREAD

**Issue by Grummel7** [PyCQA/pylint#8256]
### Bug description

# pylint: disable=missing-docstring
dict1 = {(1, 2): 3}
dict2 = {a: b for a, b in dict1}

### Command used

```shell
pylint x.py
```

### Pylint output

```shell
************* Module a
a.py:x:3: [R1721(unnecessary-comprehension), ] Unnecessary use of a comprehension, use dict(dict1) instead.
```

### Expected behavior

unnecessary-comprehension is correct, but since dict1 already is a dict, `dict(dict1)` would simply create a copy of dict1.

The correct fix here would be: `dict(dict1.keys())`

### Pylint version

```shell
pylint 2.13.9
astroid 2.11.7
Python 3.6.8 (default, Mar 18 2021, 08:58:41) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)]
```

### OS / Environment

Red Hat 8

**Comment by Pierre-Sassoulas** [comment]
``a = dict(dict1)`` is not equivalent to ``a = dict1``, one is creating another object:

```python
d = {(1, 2): 3}
e = {a: b for a, b in d}
f = dict(d)
g = d
print(id(d))
print(id(e))
print(id(f))
print(id(g))
```

```
4864 # d
5248 # another object
4992 # yet another object
4864 # g is the same object than d 
```

**Comment by clavedeluna** [comment]
It's true that `dict(dict)` isn't creating a copy, but it's creating a new dict. However, I also agree that the suggestion of `use dict(dict1) instead.` is wrong.

```
>>> dict1 = {(1, 2): 3}
>>> dict2 = {a: b for a, b in dict1}
>>> dict2
{1: 2}
```
so user wants to iterate over the key tuple. The suggestion of `dict(dict1)`
results in
```
>>> dict(dict1)
{(1, 2): 3}
```
which isn't what user wants (dict2).

`dict(dict1.keys())` is indeed the right fix here
```
>>> dict(dict1.keys())
{1: 2}
```

This is such an edge case though. How likely is anyone ever to use an iterator for a dictionary key 🤷‍♀️ 

Fix ideas:
1. easy:  only emit `unnecessary-comprehension` message if the dict has const keys. Or emit the message but don't emit a suggestion
2. less easy:  figure out a correct suggestion for this edge case and any others (other iterator as keys, iterators as values, etc)

**Comment by Grummel7** [comment]
> This is such an edge case though. How likely is anyone ever to use an iterator for a dictionary key 🤷‍♀️

I don't know why the original developer has put the second value into the key instead of the value. I just need a dict from value 1 to value 2 and then stumbled upon this wrong suggestion.

> Fix ideas:
> 
>     1. easy:  only emit `unnecessary-comprehension` message if the dict has const keys. Or emit the message but don't emit a suggestion

I don't understand that suggestion. The keys of dict1 are tuples, so they are const.

The problem here is that dict1 is not an iterator, it is only converted to an iterator when used in a place where an iterator is needed.

The suggestion to replace `{a: b for a, b in x}` with `dict(x)` is correct for all cases where x is an iterator. I think that in fact it is correct for all cases where x is not a dict.
